### DISCUSSION TITLE
Sending custom objects to button handler and using them.

### DISCUSSION THREAD

**Discussion by JayKobe6k** [DSharpPlus/DSharpPlus#1710] [Support] [answered]
Hi,
I'm having trouble understanding how to handle type of scenario where I make an API or DB call in slash command and click a button depending on what I wanna choose/do with it. At the moment I'm creating handler in slash command:

```
        client.ComponentInteractionCreated += async (sender, e) =>
            await BtnHandler(sender, e, customList);
```

```
    private async Task BtnHandler(DiscordClient sender, ComponentInteractionCreateEventArgs e, IEnumerable<Object> customList)
    {
            int selectedOption = int.Parse(e.Id.Substring("option_".Length));
        
            var selectedObject = customList.ElementAtOrDefault(selectedOption - 1);

            if (selectedObject != null)
            {
                await e.Interaction.CreateResponseAsync(InteractionResponseType.UpdateMessage,
                    new DiscordInteractionResponseBuilder().WithContent($"You selected option {selectedOption}! Here are the details:\n{selectedObject .Name} (ID: {selectedObject .Id})\n{selectedObject .Description}"));
            }
            
    }
```

But this very quickly doesn't go well because second time onwards I call slash command and click a button I'm getting not found or bad response exceptions. Because I'm making multiple handlers. From what i've read #1386 handler must come in main file where the client is actually initialized. (You guys should include such crucial information in your documentation as I was living under assumption button handler comes where buttons go) But I don't understand how to handle any type of scenario that requires any data movement between slash command module and that handler in main file.

This might seem like programming question but I feel like the library is limiting me quite a bit. So my question is how to handle what I've shown in my code, which is wrong because creates multiple handlers. But with just main file handler for button.

**Comment by Plerx2493** [comment, marked as answer]
There are several ways to solve your problem:
1. You put some kind of id in your button id and use a injected service which will hold your state
2. Designe your buttons to be stateless. Put your needed data into the custom id. For example "ban-{id}" which will always ban a certain member given some checks dont fail.
3. Use `DSharpPlus.Interactivity` and use methods like `DiscordMessage#WaitForButtonAsync`. Find the docs [here](https://dsharpplus.github.io/DSharpPlus/articles/interactivity.html)

About your feedback about the lib:
We use standard C# events and target user with an intermediate understanding of the language which should include knowledge about handling events and handling state in an threadsafe and application wide manner

**Reply by JayKobe6k** [reply]
Option 3 seemed like the only feasible option when handling list of custom objects, with multiple fields etc. So, in my main file I registered button handler 

`discord.ComponentInteractionCreated += ButtonHandler; `

which fires this empty Task:
```
        private static async Task ButtonHandler(DiscordClient sender, ComponentInteractionCreateEventArgs args)
        {
            
        }
```
So I guess the only reason it exists is for acknowledgment of interaction?


Also             `discord.UseInteractivity();` in main method to use Interactivity.

As for the command 
```

        var message = new DiscordWebhookBuilder()
            .WithContent(resultMessage)
            .AddComponents(row);

          var response=  await ctx.EditResponseAsync(message);

          var result = response.WaitForButtonAsync();
          
          int selectedOption = int.Parse(result.Result.Result.Id.Substring("option_".Length));
        
          var selectedObject = objectList.ElementAtOrDefault(selectedOption - 1);

          var newMessage = new DiscordWebhookBuilder()
              .WithContent($"You selected option {selectedOption}! Here are the details:\n{selectedObject.Title} (ID: {selectedObject.Id})\n{selectedObject.Description}");
          
          var newResponse=  await ctx.EditResponseAsync(newMessage);
```

Does that looks fine? Because it looks a bit odd, but it works without any errors in console or discord just as I wanted.

**Reply by Plerx2493** [reply]
If you use interactivity you dont need the event handler in your main. Also have in mind that your interactivity result can be a timeout

**Reply by JayKobe6k** [reply]
In what situations it can be a timeout?

**Reply by Plerx2493** [reply]
Interactivity waits only a certain timespan for an interaction. Its explained in the linked docs

**Reply by JayKobe6k** [reply]
Ahh, I was wondering what the time is but that's good that we can set to what we want. Thanks, interactivity should be ok for my use cases then
